libraries/AP_SafeRTL/AP_SafeRTL.h

```diff
@@ -0,0 +1,205 @@
+    // path_points_count is _path_points_count but passed into avoid having to take the semaphore
+    void reset_simplification(uint16_t path_points_count);
+
+    // reset pruning algorithm so that it will re-check all points in the path
+    // should be called if the existing path is altered for example when a loop as been removed
+    // path_points_count is _path_points_count but passed into avoid having to take the semaphore
+    void reset_pruning(uint16_t path_points_count);
+
+    // set all points that can be removed to zero
+    void zero_points_by_simplify_bitmask();
+    void zero_points_by_loops(uint16_t points_to_delete);
+
+    // remove all zero points from the path
+    void remove_empty_points();
+
+public:
+    // dist_point holds the closest distance reached between 2 line segments, and the point exactly between them
+    typedef struct {
+        float distance;
+        Vector3f midpoint;
+    } dist_point;
+
+    // get the closest distance between 2 line segments and the point midway between the closest points
+    static dist_point segment_segment_dist(const Vector3f& p1, const Vector3f& p2, const Vector3f& p3, const Vector3f& p4);
+private:
+
+    // de-activate SafeRTL, send warning to GCS and log to dataflash
+    void deactivate(SRTL_Actions action, const char *reason);
+
+    // logging
+    void log_action(SRTL_Actions action, const Vector3f &point = Vector3f());
+
+    // external references
+    const AP_AHRS& _ahrs;
+
+    // parameters
+    AP_Float _accuracy;
+    AP_Int16 _points_max;
+
+    // SafeRTL State Variables
+    bool _active;       // true if safeRTL is usable.  may become unusable if the path becomes too long to keep in memory, and too convoluted to be cleaned up, SafeRTL will be permanently deactivated (for the remainder of the flight)
+    bool _example_mode; // true when being called from the example sketch, logging and background tasks are disabled
+    uint32_t _last_good_position_ms;    // the last system time a last good position was reported. If no position is available for a while, SafeRTL will be disabled.
+    uint32_t _last_position_save_ms;    // the system time a position was saved to the path (used for timeout)
+    uint32_t _thorough_clean_request_ms;// the last system time the thorough cleanup was requested (set by thorough_cleanup method, used by background cleanup)
+    uint32_t _thorough_clean_complete_ms; // set to _thorough_clean_request_ms when the background thread completes the thorough cleanup
+    ThoroughCleanupType _thorough_clean_type;   // used by example sketch to test simplify and prune separately
+
+    // path variables
+    Vector3f* _path;    // points are stored in meters from EKF origin in NED
+    uint16_t _path_points_max;  // after the array has been allocated, we will need to know how big it is. We can't use the parameter, because a user could change the parameter in-flight
```

Comment:
Worth noting that a AP_Int16 means that this is still effectively just int16_t, but if uint makes the code better then it's still worth it.